<a href="https://colab.research.google.com/github/AEinstein86/OyatsiA/blob/main/DSA_8203_Group_7_colab_Project_Analysis_of_Student_Performance_Dataset_Jan_2023.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# Importing Libraries
%matplotlib inline
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from pandas_profiling import ProfileReport

In [2]:
url= 'https://raw.githubusercontent.com/AEinstein86/OyatsiA/main/StudentsPerformance.csv'
Sp = pd.read_csv(url)
# Dataset is now stored in a Pandas Dataframe

In [3]:
Sp.head(6)

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   
5  female        group B          associate's degree      standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75  
5                    none          71             83             78

In [4]:
#Pandas Profiling.
#profile=ProfileReport(Sp,title="Pandas Profiling Report",explorative=True)

In [5]:
#profile.to_widgets()

In [6]:
#profile.to_file("Exploratory_Data_Analysis.html")

In [7]:
# Dataset Shape
Sp.shape

(1000, 8)

In [8]:
# Summary Statistics 
Sp.describe()

math score  reading score  writing score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

In [9]:
#Dataset Columns
Sp.columns

Index(['gender', 'race/ethnicity', 'parental level of education', 'lunch',
       'test preparation course', 'math score', 'reading score',
       'writing score'],
      dtype='object')

In [10]:
# Find Unique Values
Sp.nunique()

gender                          2
race/ethnicity                  5
parental level of education     6
lunch                           2
test preparation course         2
math score                     81
reading score                  72
writing score                  77
dtype: int64

In [11]:
#Cleaning the data.No missing values in the dataset.
Sp.isna().sum()

gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

In [12]:
Sp['total_score'] = (Sp['math score'] + Sp['reading score'] +Sp['writing score'])/3
Sp.head(2)

gender race/ethnicity parental level of education     lunch  \
0  female        group B           bachelor's degree  standard   
1  female        group C                some college  standard   

  test preparation course  math score  reading score  writing score  \
0                    none          72             72             74   
1               completed          69             90             88   

   total_score  
0    72.666667  
1    82.333333

In [13]:
# Drop columns(race/ethnicity,parental level of education)that are not necessary for my evaluation.
Student=Sp.drop(["race/ethnicity","parental level of education"],axis=1)

In [14]:
Student

gender         lunch test preparation course  math score  reading score  \
0    female      standard                    none          72             72   
1    female      standard               completed          69             90   
2    female      standard                    none          90             95   
3      male  free/reduced                    none          47             57   
4      male      standard                    none          76             78   
..      ...           ...                     ...         ...            ...   
995  female      standard               completed          88             99   
996    male  free/reduced                    none          62             55   
997  female  free/reduced               completed          59             71   
998  female      standard               completed          68             78   
999  female  free/reduced                    none          77             86   

     writing score  total_score  
0               74    72.666667  
1               88    82.333333  
2               93    92.666667  
3               44    49.333333  
4               75    76.333333  
..             ...          ...  
995             95    94.000000  
996             55    57.333333  
997             65    65.000000  
998             77    74.333333  
999             86    83.000000  

[1000 rows x 7 columns]

Relationship Analysis

In [15]:
corelation = Student.corr()

In [16]:
#There is a strong correlation between reading score and writing score.Math score and reading score. 
#Example of color maps:plasma,inferno,magma,cividis,viridis.
plt.figure(figsize=(10,8))
sns.heatmap(corelation,xticklabels=corelation.columns,yticklabels=corelation.columns,annot=True,cmap="plasma")
plt.savefig("Heatmap.png")

In [17]:
#Pair plot
sns.pairplot(Student,hue="gender")
plt.savefig("PairPlot.png")

In [18]:
#relplot
sns.relplot(x="math score",y="reading score",hue="gender",data=Student)
plt.savefig('Scatter Plot.png')

In [19]:
#Histogram:Math Score Distribution.
#Can be deduced that most people in the dataset are getting a math score of 66.089.
#Student["math score"].mean()=66.089
#most distribution is between 60 to 80 column.
plt.figure(figsize=(10,8))
plt.ylabel("Frequency")
plt.title("Math Score Distribution")
sns.distplot(Student["math score"],color="r",bins=5)
plt.savefig("Math Score Distribution.png")

/usr/local/lib/python3.8/dist-packages/seaborn/distributions.py:2619: FutureWarning: `distplot` is a deprecated function and will be removed in a future version. Please adapt your code to use either `displot` (a figure-level function with similar flexibility) or `histplot` (an axes-level function for histograms).
  warnings.warn(msg, FutureWarning)


In [20]:
#Histogram (Reading Score Distribution).
#Can be deduced that most people in the dataset are getting a math score of 66.089.
#Student["reading score"].mean()=66.089
plt.figure(figsize=(10,8))
sns.distplot(Student["reading score"],color="b",bins=5)
plt.ylabel("Frequency")
plt.title("Reading Score Distribution")
plt.savefig("Reading Score Distribution.png")

/usr/local/lib/python3.8/dist-packages/seaborn/distributions.py:2619: FutureWarning: `distplot` is a deprecated function and will be removed in a future version. Please adapt your code to use either `displot` (a figure-level function with similar flexibility) or `histplot` (an axes-level function for histograms).
  warnings.warn(msg, FutureWarning)


In [21]:
#Histogram (Writing Score Distribution).
#Can be deduced that most people in the dataset are getting a math score of 66.089.
#Student["writing score"].mean()=68.054
#most distribution is between 60-80 column.
sns.distplot(Student["writing score"],color="y",bins=5)
plt.ylabel("Frequency")
plt.title("Writing Score Distribution");

/usr/local/lib/python3.8/dist-packages/seaborn/distributions.py:2619: FutureWarning: `distplot` is a deprecated function and will be removed in a future version. Please adapt your code to use either `displot` (a figure-level function with similar flexibility) or `histplot` (an axes-level function for histograms).
  warnings.warn(msg, FutureWarning)


In [22]:
#Histogram (Total Score Distribution).
#Can be deduced that most people in the dataset are getting a math score of 66.089.
#Student["reading score"].mean()=66.089
plt.figure(figsize=(10,8))
sns.distplot(Student["total_score"],color="b",bins=5)
plt.ylabel("Frequency")
plt.title("Total Score Distribution")
plt.savefig("Total Score Distribution.png")

/usr/local/lib/python3.8/dist-packages/seaborn/distributions.py:2619: FutureWarning: `distplot` is a deprecated function and will be removed in a future version. Please adapt your code to use either `displot` (a figure-level function with similar flexibility) or `histplot` (an axes-level function for histograms).
  warnings.warn(msg, FutureWarning)


In [23]:
Student["writing score"].mean()

68.054

In [24]:
print("The number of students having a maximum score in math:",len(Student.loc[Student["math score"]==100]))


The number of students having a maximum score in math: 7


In [25]:
print("The number of students having a maximum score in writing:",len(Student.loc[Student["writing score"]==100]))

The number of students having a maximum score in writing: 14


In [26]:
print("The number of students having a maximum score in reading:",len(Student.loc[Student["reading score"]==100]))


The number of students having a maximum score in reading: 17


In [27]:
#Perfect Score (Scored 100% in all three categories).
perfect_math_score=Student["math score"]==100
perfect_writing_score=Student["writing score"]==100
perfect_reading_score=Student["reading score"]==100

Perfect_Score=Student.loc[(perfect_math_score)&(perfect_writing_score)&(perfect_reading_score)]
Perfect_Score

gender     lunch test preparation course  math score  reading score  \
458  female  standard                    none         100            100   
916    male  standard               completed         100            100   
962  female  standard                    none         100            100   

     writing score  total_score  
458            100        100.0  
916            100        100.0  
962            100        100.0

In [28]:
#Number of students with a perfect score (a score of 100).
len(Perfect_Score)

3

In [29]:
#Math Score (According to gender).
print('----Females----')
print('Max. math Score: ', Student[Student['gender'] == 'female']['math score'].max())
print('Min. math Score: ', Student[Student['gender'] == 'female']['math score'].min())
print('Average math Score: ', Student[Student['gender'] == 'female']['math score'].mean())
print('----Males----')
print('Max. math Score: ', Student[Student['gender'] == 'male']['math score'].max())
print('Min. math Score: ', Student[Student['gender'] == 'male']['math score'].min())
print('Average math Score: ', Student[Student['gender'] == 'male']['math score'].mean())


----Females----
Max. math Score:  100
Min. math Score:  0
Average math Score:  63.633204633204635
----Males----
Max. math Score:  100
Min. math Score:  27
Average math Score:  68.72821576763485


In [30]:
#Writing Score (According to gender).
print('----Females----')
print('Max. writing Score: ', Student[Student['gender'] == 'female']['writing score'].max())
print('Min. writing Score: ', Student[Student['gender'] == 'female']['writing score'].min())
print('Average writing Score: ', Student[Student['gender'] == 'female']['writing score'].mean())
print('----Males----')
print('Max. writing Score: ', Student[Student['gender'] == 'male']['writing score'].max())
print('Min. writing Score: ', Student[Student['gender'] == 'male']['writing score'].min())
print('Average writing Score: ', Student[Student['gender'] == 'male']['writing score'].mean())


----Females----
Max. writing Score:  100
Min. writing Score:  10
Average writing Score:  72.46718146718146
----Males----
Max. writing Score:  100
Min. writing Score:  15
Average writing Score:  63.31120331950208


In [31]:
#Reading Score (According to gender).
print('----Females----')
print('Max. reading Score: ', Student[Student['gender'] == 'female']['reading score'].max())
print('Min. reading Score: ', Student[Student['gender'] == 'female']['reading score'].min())
print('Average reading Score: ', Student[Student['gender'] == 'female']['reading score'].mean())
print('----Males----')
print('Max. reading Score: ', Student[Student['gender'] == 'male']['reading score'].max())
print('Min. reading Score: ', Student[Student['gender'] == 'male']['reading score'].min())
print('Average reading Score: ', Student[Student['gender'] == 'male']['reading score'].mean())


----Females----
Max. reading Score:  100
Min. reading Score:  17
Average reading Score:  72.60810810810811
----Males----
Max. reading Score:  100
Min. reading Score:  23
Average reading Score:  65.47302904564316


In [32]:
#Box Plots
plt.figure(figsize=(12,5))

plt.subplot(1,3,1)
sns.boxplot(x = 'gender', y = 'math score', data = Student,palette = ['coral', 'lawngreen'])

plt.subplot(1,3,2)
sns.boxplot(x = 'gender', y = 'reading score', data = Student,palette = ['coral', 'lawngreen'])

plt.subplot(1,3,3)
sns.boxplot(x = 'gender', y = 'writing score', data = Student,palette = ['coral', 'lawngreen'])

plt.tight_layout()
plt.savefig("All categories.png")

In [33]:
#Math Score Box Plot according to gender.
plt.figure(figsize=(10,8))
sns.catplot(x="math score",kind="box",col="gender",data=Student)
plt.savefig("Box Plot.png")


In [34]:
sns.countplot(x="test preparation course",hue="gender",data=Student,palette="rainbow",saturation=1,edgecolor=(0,0,0),linewidth=2)

In [35]:
sns.countplot(x="lunch",hue="gender",data=Student,palette="rainbow",saturation=1,edgecolor=(0,0,0),linewidth=2)

In [36]:
#Gender Count
sns.set(font_scale=1.25)
sns.set_style('whitegrid')
splot=sns.countplot(x="gender",hue="gender",data=Student,palette="rainbow",saturation=1,edgecolor=(0,0,0),linewidth=2)
for p in splot.patches:
    splot.annotate(format(p.get_height(), '.1f'), 
                   (p.get_x() + p.get_width() / 2., p.get_height()), 
                   ha = 'center', va = 'center', 
                   size=15,
                   xytext = (0, 9), 
                   textcoords = 'offset points')
plt.xlabel("Gender", size=14)
plt.ylabel("Count", size=14)
plt.legend([])
plt.savefig("Gender Count.png")

In [37]:
#Point Plots.
plt.figure(figsize=(14,4))
plt.subplot(1,3,1)
sns.pointplot(x = "gender", y = "math score",data= Student)

plt.subplot(1,3,2)
sns.pointplot(x="gender",y="reading score",data=Student)

plt.subplot(1,3,3)
sns.pointplot(x="gender",y="writing score",data=Student)

plt.tight_layout()
plt.savefig("Point Plot.png")

In [38]:
#Bar Plots.
plt.figure(figsize=(10,4))

plt.subplot(1,3,1)
splot=sns.barplot(x = 'gender', y = 'reading score', data = Student)
for p in splot.patches:
    splot.annotate(format(p.get_height(), '.1f'), 
                   (p.get_x() + p.get_width() / 2., p.get_height()), 
                   ha = 'center', va = 'center', 
                   size=15,
                   xytext = (0, 9), 
                   textcoords = 'offset points')
plt.xlabel("Gender", size=14)
plt.ylabel("Reading Score", size=14)

plt.subplot(1,3,2)
splot=sns.barplot(x = 'gender', y = 'writing score', data = Student)
for p in splot.patches:
    splot.annotate(format(p.get_height(), '.1f'), 
                   (p.get_x() + p.get_width() / 2., p.get_height()), 
                   ha = 'center', va = 'center', 
                   size=15,
                   xytext = (0, 9), 
                   textcoords = 'offset points')
plt.xlabel("Gender", size=14)
plt.ylabel("Writing Score", size=14)

plt.subplot(1,3,3)
splot=sns.barplot(x = 'gender', y = 'math score', data = Student)
for p in splot.patches:
    splot.annotate(format(p.get_height(), '.1f'), 
                   (p.get_x() + p.get_width() / 2., p.get_height()), 
                   ha = 'center', va = 'center', 
                   size=15,
                   xytext = (0, 9), 
                   textcoords = 'offset points')
plt.xlabel("Gender", size=14)
plt.ylabel("Math Score", size=14)

plt.tight_layout()
plt.savefig("Bar Plot Comparison")

In [39]:
#Violin Plots.
plt.figure(figsize=(14,4))
plt.subplot(1,3,1)
ax=sns.violinplot(x="gender",y="reading score",hue="gender",data=Student)
plt.subplot(1,3,2)
ax=sns.violinplot(x="gender",y="math score",hue="gender",data=Student)
plt.subplot(1,3,3)
ax=sns.violinplot(x="gender",y="writing score",hue="gender",data=Student)
plt.tight_layout()
plt.savefig("Gender violin Plot")

In [40]:
#Bar Plots acccording to Lunch.
plt.figure(figsize=(14,4))

plt.subplot(1,3,1)
splot=sns.barplot(x = 'lunch', y = 'reading score', hue = 'gender', data = Student)
for p in splot.patches:
    splot.annotate(format(p.get_height(), '.1f'), 
                   (p.get_x() + p.get_width() / 2., p.get_height()), 
                   ha = 'center', va = 'center', 
                   size=15,
                   xytext = (0, 9), 
                   textcoords = 'offset points')
plt.xlabel("Lunch", size=14)
plt.ylabel("Reading Score", size=14)

plt.subplot(1,3,2)
splot=sns.barplot(x = 'lunch', y = 'writing score',hue = 'gender', data = Student)
for p in splot.patches:
    splot.annotate(format(p.get_height(), '.1f'), 
                   (p.get_x() + p.get_width() / 2., p.get_height()), 
                   ha = 'center', va = 'center', 
                   size=15,
                   xytext = (0, 9), 
                   textcoords = 'offset points')
plt.xlabel("Lunch", size=14)
plt.ylabel("Writing Score", size=14)

plt.subplot(1,3,3)
splot=sns.barplot(x = 'lunch', y = 'math score',hue = 'gender', data = Student)
for p in splot.patches:
    splot.annotate(format(p.get_height(), '.1f'), 
                   (p.get_x() + p.get_width() / 2., p.get_height()), 
                   ha = 'center', va = 'center', 
                   size=15,
                   xytext = (0, 9), 
                   textcoords = 'offset points')
plt.xlabel("Lunch", size=14)
plt.ylabel("Math Score", size=14)

plt.tight_layout()
plt.savefig("Lunch.png")

In [41]:
#Bar Plots according to Test Preparation Score.
plt.figure(figsize=(14,4))

plt.subplot(1,3,1)
splot=sns.barplot(x = 'test preparation course', y = 'reading score', hue = 'gender', data = Student)
for p in splot.patches:
    splot.annotate(format(p.get_height(), '.1f'), 
                   (p.get_x() + p.get_width() / 2., p.get_height()), 
                   ha = 'center', va = 'center', 
                   size=15,
                   xytext = (0, 9), 
                   textcoords = 'offset points')
plt.xlabel("Test Preparation Course", size=14)
plt.ylabel("Reading Score", size=14)

plt.subplot(1,3,2)
splot=sns.barplot(x = 'test preparation course', y = 'writing score',hue = 'gender', data = Student)
for p in splot.patches:
    splot.annotate(format(p.get_height(), '.1f'), 
                   (p.get_x() + p.get_width() / 2., p.get_height()), 
                   ha = 'center', va = 'center', 
                   size=15,
                   xytext = (0, 9), 
                   textcoords = 'offset points')
plt.xlabel("Test Preparation Course", size=14)
plt.ylabel("Writing Score", size=14)

plt.subplot(1,3,3)
splot=sns.barplot(x = 'test preparation course', y = 'math score',hue = 'gender', data = Student)
for p in splot.patches:
    splot.annotate(format(p.get_height(), '.1f'), 
                   (p.get_x() + p.get_width() / 2., p.get_height()), 
                   ha = 'center', va = 'center', 
                   size=15,
                   xytext = (0, 9), 
                   textcoords = 'offset points')
plt.xlabel("Test Preparation Course", size=14)
plt.ylabel("Math Score", size=14)

plt.tight_layout()
plt.savefig("Test Preparation Course.png")

In [42]:
passing_marks = 40
Student['Pass_Status'] = np.where(Student['total_score']<passing_marks,'F','P')
Student['Pass_Status'].value_counts()

P    970
F     30
Name: Pass_Status, dtype: int64

In [43]:
Student.head(2)

gender     lunch test preparation course  math score  reading score  \
0  female  standard                    none          72             72   
1  female  standard               completed          69             90   

   writing score  total_score Pass_Status  
0             74    72.666667           P  
1             88    82.333333           P

In [44]:
#Assigning Grades.
# Function to assign grades.

def get_grade(marks):
    if marks >= 82 and marks <= 100:
        return 'A+'
    elif marks >=73 and marks < 82:
        return 'A'
    elif marks >=64 and marks < 73:
        return 'B+'
    elif marks >= 55 and marks < 64:
        return 'B'
    elif marks >=46 and marks < 55:
        return 'C'
    elif marks >= 35 and marks < 46:
        return 'P'
    elif marks < 35:
        return 'F'

In [45]:
#Assigning Grades Continuation.
Student['Reading Grade'] = Student['reading score'].apply(get_grade)
Student['Writing Grade'] = Student['writing score'].apply(get_grade)
Student['Math Grade'] = Student['math score'].apply(get_grade)

In [46]:
def grades(pass_status,total_score):
    if pass_status == 'F':
        return 'F'
    if total_score >=90:
        return 'A+'
    if total_score >=80:
        return 'A'
    if total_score >=70:
        return 'B'
    if total_score >=60:
        return 'C'
    if total_score >=50:
        return 'D'
    if total_score >=40:
        return 'E'

In [47]:
Student['Grade'] =Student.apply(lambda x : grades(x['Pass_Status'] ,x['total_score']) ,axis =1)
Student.head(2)

gender     lunch test preparation course  math score  reading score  \
0  female  standard                    none          72             72   
1  female  standard               completed          69             90   

   writing score  total_score Pass_Status Reading Grade Writing Grade  \
0             74    72.666667           P            B+             A   
1             88    82.333333           P            A+            A+   

  Math Grade Grade  
0         B+     B  
1         B+     A

In [48]:
#Assigning Grades Continuation.
sns.set_style('whitegrid')
plt.figure(figsize=(16,5))
plt.subplot(1,3,1)
splot=sns.countplot(x ='Math Grade', data = Student,order = ['A+','A','B+','B','C','P','F'],palette='deep')
for p in splot.patches:
    splot.annotate(format(p.get_height(), '.1f'), 
                   (p.get_x() + p.get_width() / 2., p.get_height()), 
                   ha = 'center', va = 'center', 
                   size=15,
                   xytext = (0, 9), 
                   textcoords = 'offset points')
plt.subplot(1,3,2)
splot=sns.countplot(x ='Reading Grade', data = Student,order = ['A+','A','B+','B','C','P','F'],palette='deep')
plt.subplot(1,3,1)
for p in splot.patches:
    splot.annotate(format(p.get_height(), '.1f'), 
                   (p.get_x() + p.get_width() / 2., p.get_height()), 
                   ha = 'center', va = 'center', 
                   size=15,
                   xytext = (0, 9), 
                   textcoords = 'offset points')
plt.subplot(1,3,3)
splot=sns.countplot(x ='Writing Grade', data = Student,order = ['A+','A','B+','B','C','P','F'],palette='deep')
for p in splot.patches:
    splot.annotate(format(p.get_height(), '.1f'), 
                   (p.get_x() + p.get_width() / 2., p.get_height()), 
                   ha = 'center', va = 'center', 
                   size=15,
                   xytext = (0, 9), 
                   textcoords = 'offset points')
plt.tight_layout()
plt.savefig("Grade Count Plot.png")


<ipython-input-48-7f69edbea618>:15: MatplotlibDeprecationWarning: Adding an axes using the same arguments as a previous axes currently reuses the earlier instance.  In a future version, a new instance will always be created and returned.  Meanwhile, this warning can be suppressed, and the future behavior ensured, by passing a unique label to each axes instance.
  plt.subplot(1,3,1)


In [49]:
#Establishing the number of students who attained a perfect grade.
perfect_writing = Student['Writing Grade'] == 'A+'
perfect_reading = Student['Reading Grade'] == 'A+'
perfect_math = Student['Math Grade'] == 'A+'

perfect_grade = Student[(perfect_math) & (perfect_reading) & (perfect_writing)]
print('Number of students having maximum grade(A+) in all three subjects: ',len(perfect_grade))


Number of students having maximum grade(A+) in all three subjects:  95


In [50]:
#Establishing the number of students who attained the least grade in all three categries.
minimum_math = Student['Math Grade'] == 'F'
minimum_reading = Student['Reading Grade'] == 'F'
minimum_writing = Student['Writing Grade'] == 'F'



minimum_grade = Student[(minimum_math) & (minimum_reading) & (minimum_writing)]
print('Number of students having minimum grade(F) in all three subjects: ',len(minimum_grade))


Number of students having minimum grade(F) in all three subjects:  8


In [51]:
#Failed Students
failed_students = Student[(minimum_math) | (minimum_reading)|(minimum_writing)]
failed = len(failed_students)
print('Total Number of students who failed are: {}' .format(len(failed_students)))


Total Number of students who failed are: 29


In [52]:
#Passed Students
passed_students = len(Student) - len(failed_students)
print('Total Number of students who passed are: {}' .format(passed_students))


Total Number of students who passed are: 971


In [53]:
#Pie chart to illustrate all the data above (passed students and failed students).
plt.figure(figsize=(8,6))

#Data to plot
labels = 'Passed Students', 'Failed Students'
sizes = [passed_students,failed]
colors = ['b','r']
explode = (0.32,0.0)

#Plot
plt.pie(sizes,explode = explode, labels = labels,colors = colors,
       autopct='%.2f%%',shadow = True, startangle=360)
plt.axis('equal')

plt.legend()
plt.savefig("piechart.png")
plt.show()

Linear Regression Modelling Using Scikit Learn.

In [54]:
from sklearn.linear_model import LogisticRegression

In [55]:
from sklearn import linear_model

In [56]:
X_shaped=Student["math score"].values

In [57]:
X_shaped=Student["math score"].values.reshape(-1,1)
X_shaped

array([[ 72],
       [ 69],
       [ 90],
       [ 47],
       [ 76],
       [ 71],
       [ 88],
       [ 40],
       [ 64],
       [ 38],
       [ 58],
       [ 40],
       [ 65],
       [ 78],
       [ 50],
       [ 69],
       [ 88],
       [ 18],
       [ 46],
       [ 54],
       [ 66],
       [ 65],
       [ 44],
       [ 69],
       [ 74],
       [ 73],
       [ 69],
       [ 67],
       [ 70],
       [ 62],
       [ 69],
       [ 63],
       [ 56],
       [ 40],
       [ 97],
       [ 81],
       [ 74],
       [ 50],
       [ 75],
       [ 57],
       [ 55],
       [ 58],
       [ 53],
       [ 59],
       [ 50],
       [ 65],
       [ 55],
       [ 66],
       [ 57],
       [ 82],
       [ 53],
       [ 77],
       [ 53],
       [ 88],
       [ 71],
       [ 33],
       [ 82],
       [ 52],
       [ 58],
       [  0],
       [ 79],
       [ 39],
       [ 62],
       [ 69],
       [ 59],
       [ 67],
       [ 45],
       [ 60],
       [ 61],
       [ 39],
       [ 58],
      

In [58]:
Y_shaped=Student["total_score"].values.reshape(-1,1)
Y_shaped

array([[ 72.66666667],
       [ 82.33333333],
       [ 92.66666667],
       [ 49.33333333],
       [ 76.33333333],
       [ 77.33333333],
       [ 91.66666667],
       [ 40.66666667],
       [ 65.        ],
       [ 49.33333333],
       [ 54.66666667],
       [ 45.        ],
       [ 73.        ],
       [ 73.33333333],
       [ 53.66666667],
       [ 74.        ],
       [ 87.66666667],
       [ 26.        ],
       [ 44.66666667],
       [ 57.66666667],
       [ 66.        ],
       [ 70.        ],
       [ 50.33333333],
       [ 71.66666667],
       [ 75.        ],
       [ 73.        ],
       [ 59.33333333],
       [ 70.33333333],
       [ 68.33333333],
       [ 69.        ],
       [ 72.33333333],
       [ 63.        ],
       [ 64.33333333],
       [ 40.        ],
       [ 88.66666667],
       [ 80.33333333],
       [ 79.33333333],
       [ 57.66666667],
       [ 84.33333333],
       [ 56.66666667],
       [ 56.66666667],
       [ 66.33333333],
       [ 58.66666667],
       [ 63

In [59]:
#Our next step is to divide the data into attributes and labels.
#Attributes are the independent variables while labels are dependent variables whose values are to be predicted.
#We want to predict the total_score depending upon the math score recorded. 
X_shaped=Student["math score"].values.reshape(-1,1)
Y_shaped=Student["total_score"].values.reshape(-1,1)

In [60]:
#random state can be any number.
#Next, we split 80% of the data to the training set while 20% of the data to test set using below code.
#The test_size variable is where we actually specify the proportion of the test set.
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(X_shaped,Y_shaped,test_size=0.2,random_state=0)
print(x_train.shape)
print(x_test.shape)
print(y_train)
print(y_test)

(800, 1)
(200, 1)
[[ 76.        ]
 [ 78.33333333]
 [ 57.        ]
 [ 93.33333333]
 [ 69.        ]
 [ 50.        ]
 [ 78.33333333]
 [ 82.33333333]
 [ 74.        ]
 [ 70.        ]
 [ 54.33333333]
 [ 91.66666667]
 [ 57.66666667]
 [ 63.33333333]
 [ 34.66666667]
 [ 69.        ]
 [ 78.        ]
 [ 63.33333333]
 [ 56.66666667]
 [ 72.33333333]
 [ 76.66666667]
 [ 66.66666667]
 [ 62.33333333]
 [ 73.        ]
 [ 64.33333333]
 [ 78.33333333]
 [ 87.66666667]
 [ 61.33333333]
 [ 44.66666667]
 [ 63.66666667]
 [ 44.        ]
 [ 63.66666667]
 [ 53.66666667]
 [ 48.33333333]
 [ 59.33333333]
 [ 66.        ]
 [ 71.        ]
 [ 67.        ]
 [ 69.33333333]
 [ 58.66666667]
 [ 77.33333333]
 [ 51.66666667]
 [ 79.        ]
 [ 59.        ]
 [ 75.        ]
 [ 72.33333333]
 [ 61.66666667]
 [ 92.66666667]
 [ 63.        ]
 [ 84.66666667]
 [ 77.33333333]
 [ 83.        ]
 [ 78.33333333]
 [ 79.66666667]
 [ 41.        ]
 [ 76.        ]
 [ 53.66666667]
 [ 87.33333333]
 [ 68.66666667]
 [ 84.33333333]
 [ 62.33333333]
 [ 75.

Initializing the model

In [61]:
from sklearn.linear_model import LinearRegression
lm=LinearRegression()

Fitting the model

In [62]:
#After splitting the data into training and testing sets, finally, the time is to train our algorithm.
lm.fit(x_train,y_train)

LinearRegression()

In [63]:
LinearRegression()

LinearRegression()

In [64]:
#finds the best value for the intercept.
Intercept=lm.intercept_
print(f"the y-intercept is:{Intercept}")


the y-intercept is:[10.64277615]


In [65]:
#finds the best value for the Coefficient.
Coefficient=lm.coef_
print(f"the Coefficient is:{Coefficient}")


the Coefficient is:[[0.8657433]]


This means that for every one unit of change in math score, the change in the total_score is about 0.85%.

Now that we have trained our algorithm, it’s time to make some predictions. To do so, we will use our test data and see how accurately our algorithm predicts the percentage score.

In [66]:
# making the predictions using the predict function.
y_pred = lm.predict(x_test)


In [67]:
type(y_pred)

numpy.ndarray

Lets make a comparison between the actual values for x_test and the predicted values.

In [68]:
comparison_df = pd.DataFrame({'Actual':y_test.flatten(),'Predicted':y_pred.flatten()})
comparison_df

Actual  Predicted
0    69.333333  64.318861
1    77.333333  85.962443
2    45.333333  45.272508
3    67.666667  77.305010
4    74.333333  70.379064
..         ...        ...
195  61.333333  65.184604
196  84.000000  74.707780
197  80.666667  83.365213
198  69.666667  66.916091
199  66.000000  64.318861

[200 rows x 2 columns]

Visualizing the comparison result as a bar graph. Note: As the number of records is huge, for representation purpose I’m taking just 25 records.

In [69]:
df1 = comparison_df.head(25)
df1.plot(kind='bar',figsize=(16,10))
plt.grid(which='major', linestyle='-', linewidth='0.5', color='green')
plt.grid(which='minor', linestyle=':', linewidth='0.5', color='black')
plt.show()

In [70]:
# Plotting the line of best fit.
plt.scatter(x_test, y_test,  color='red')
plt.plot(x_test, y_pred, color='blue', linewidth=5)
plt.ylabel("Predicted total_scores",fontsize=15)
plt.xlabel("Actual Math Scores",fontsize=15)
plt.title("Predicted Total_Scores Vs Actual Math_Scores")
plt.savefig("mini-project.png")
plt.show()

The straight line in the above graph shows our algorithm is correct.

The final step is to evaluate the performance of the algorithm. This step is particularly important to compare how well different algorithms perform on a particular dataset. For regression algorithms, three evaluation metrics are commonly used

In [71]:
from sklearn import metrics
print('Mean Absolute Error:', metrics.mean_absolute_error(y_test, y_pred))  
print('Mean Squared Error:', metrics.mean_squared_error(y_test, y_pred))  
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_pred)))


Mean Absolute Error: 4.950906780017983
Mean Squared Error: 36.13800474366885
Root Mean Squared Error: 6.011489394789685


You can see that the value of root mean squared error is 6.01, which is more than 10% of the mean value of the percentages of all the scores. This means that our algorithm was not very accurate but can still make reasonably good predictions.